checks.d/docker_daemon.py

```diff
@@ -0,0 +1,640 @@
+    "image_name": lambda c: image_tag_extractor(c, 0),
+    "image_tag": lambda c: image_tag_extractor(c, 1),
+    "container_command": lambda c: [c["Command"]],
+    "container_name": lambda c: [str(c['Names'][0]).lstrip("/")] if c["Names"] else [c['Id'][:11]],
+}
+
+CONTAINER = "container"
+PERFORMANCE = "performance"
+FILTERED = "filtered"
+IMAGE = "image"
+
+
+def get_mountpoints(docker_root):
+    mountpoints = {}
+    for metric in CGROUP_METRICS:
+        mountpoints[metric["cgroup"]] = find_cgroup(metric["cgroup"], docker_root)
+    return mountpoints
+
+def get_filters(include, exclude):
+    # The reasoning is to check exclude first, so we can skip if there is no exclude
+    if not exclude:
+        return
+
+    filtered_tag_names = []
+    exclude_patterns = []
+    include_patterns = []
+
+    # Compile regex
+    for rule in exclude:
+        exclude_patterns.append(re.compile(rule))
+        filtered_tag_names.append(rule.split(':')[0])
+    for rule in include:
+        include_patterns.append(re.compile(rule))
+        filtered_tag_names.append(rule.split(':')[0])
+
+    return set(exclude_patterns), set(include_patterns), set(filtered_tag_names)
+
+
+class DockerDaemon(AgentCheck):
+    """Collect metrics and events from Docker API and cgroups."""
+
+    def __init__(self, name, init_config, agentConfig, instances=None):
+        if instances is not None and len(instances) > 1:
+            raise Exception("Docker check only supports one configured instance.")
+        AgentCheck.__init__(self, name, init_config,
+                            agentConfig, instances=instances)
+
+        self.init_success = False
+        self.init()
+
+
+    def init(self):
+        try:
+            # We configure the check with the right cgroup settings for this host
+            # Just needs to be done once
+            instance = self.instances[0]
+            set_docker_settings(self.init_config, instance)
+
+            self.client = get_client()
+            self._docker_root = self.init_config.get('docker_root', '/')
+            self._mountpoints = get_mountpoints(self._docker_root)
+            self.cgroup_listing_retries = 0
+            self._latest_size_query = 0
+
+            # At first run we'll just collect the events from the latest 60 secs
+            self._last_event_collection_ts = int(time.time()) - 60
+
+            # Set tagging options
+            self.custom_tags = instance.get("tags", [])
+            self.collect_labels_as_tags = instance.get("collect_labels_as_tags", [])
+            self.tag_names = {
+                CONTAINER: instance.get("container_tags", DEFAULT_CONTAINER_TAGS),
+                PERFORMANCE: instance.get("performance_tags", DEFAULT_PERFORMANCE_TAGS),
+                IMAGE: instance.get('image_tags', DEFAULT_IMAGE_TAGS)
+
+            }
+
+            # Set filtering settings
+            if not instance.get("exclude"):
+                self._filtering_enabled = False
```

Comment:
Since it seems `include` is meaningless without `exclude`, should we show a warning for the case `instance.get('include') and not instance.get('exclude')`? I can see that being a common mistake